### PR TITLE
docs: use `lychee` to check link URLs

### PR DESCRIPTION
**Pull request by davidhewitt**
Replaces #3939 

`lychee` is better maintained and more comprehensive than `mdbook-linkcheck`.

I was able to get it to run to check the `guide` links and also the (external) rustdoc links.

### CODE REVIEW

**Comment by davidhewitt** [comment]
Ok, this should be good to review. Fixed so many broken links 😅

File: Contributing.md [hunk 1]
```diff
@@ -70,6 +70,12 @@ First, install [`mdbook`][mdbook] and [`nox`][nox]. Then, run
 nox -s build-guide -- --open
 ```
 
+To check all links in the guide are valid, also install [`lychee`][lychee] use the `check-guide` session instead:
```
Comment (mejrs) [inline Contributing.md:73]: this sentence does not sound quite right

File: guide/src/conversions/tables.md [hunk 2]
```diff
@@ -43,7 +43,7 @@ The table below contains the Python type and the corresponding function argument
 | `typing.Sequence[T]` | `Vec<T>`                 | `&PySequence`        |
 | `typing.Mapping[K, V]` | `HashMap<K, V>`, `BTreeMap<K, V>`, `hashbrown::HashMap<K, V>`[^3], `indexmap::IndexMap<K, V>`[^4] | `&PyMapping` |
 | `typing.Iterator[Any]` | -                      | `&PyIterator`        |
-| `typing.Union[...]` | See [`#[derive(FromPyObject)]`](traits.html#deriving-a-hrefhttpsdocsrspyo3latestpyo3conversiontraitfrompyobjecthtmlfrompyobjecta-for-enums) | - |
```
Comment (LilyFoote) [inline guide/src/conversions/tables.md:46]: :scream:

**Review by mejrs — APPROVED** [review]
Thanks!

File: Contributing.md (hunk 1 above)
Comment (davidhewitt) [reply Contributing.md]: Good point, I've added "and" - "and use the..."

**Review by LilyFoote — APPROVED** [review]

**Pull request merged**